bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/WebSocketArt.java

```diff
@@ -24,140 +60,882 @@
+
+        public String getEvent() {
+            return Optional.ofNullable(event).orElse("");
+        }
+
+        public String getData() {
+            return Optional.ofNullable(data).map(a -> a.getAsString()).orElse("");
+        }
+
+        public void putBinaryData(byte[] arr, int off, int len) {
+            this.binData = new BinaryData(arr, off, len);
+        }
+
+        public BinaryData getBinaryData() {
+            return Optional.ofNullable(binData).orElse(new BinaryData(new byte[0], 0, 0));
+        }
     }
 
     @Override
-    public void onWebSocketText(@Nullable String msgarg) {
+    public void onWebSocketBinary(byte @Nullable [] arr, int off, int len) {
+        if (arr == null) {
+            return;
+        }
+        super.onWebSocketBinary(arr, off, len);
+        String msg = extractMsg(arr, off, len, true);
+        // offset is start of binary data
+        int offset = ByteBuffer.wrap(arr, off, len).getShort() + off + 2; // 2 = length of Short
+        try {
+            JSONMessage jsonMsg = remoteControllerWebSocket.gson.fromJson(msg, JSONMessage.class);
+            if (jsonMsg == null) {
+                return;
+            }
+            switch (jsonMsg.getEvent()) {
+                case "d2d_service_message":
+                    jsonMsg.putBinaryData(arr, offset, len);
+                    handleD2DServiceMessage(jsonMsg);
+                    break;
+                default:
+                    logger.debug("{}: WebSocketArt(binary) Unknown event: {}", host, msg);
+            }
+        } catch (JsonSyntaxException e) {
+            logger.warn("{}: {}: Error ({}) in message: {}", host, className, e.getMessage(), msg);
+        }
+    }
+
+    @Override
+    public synchronized void onWebSocketText(@Nullable String msgarg) {
         if (msgarg == null) {
             return;
         }
         String msg = msgarg.replace('\n', ' ');
         super.onWebSocketText(msg);
         try {
             JSONMessage jsonMsg = remoteControllerWebSocket.gson.fromJson(msg, JSONMessage.class);
-
-            switch (jsonMsg.event) {
+            if (jsonMsg == null) {
+                return;
+            }
+            switch (jsonMsg.getEvent()) {
                 case "ms.channel.connect":
-                    logger.debug("Art channel connected");
+                    logger.debug("{}: Art channel connected", host);
                     break;
                 case "ms.channel.ready":
-                    logger.debug("Art channel ready");
+                    logger.debug("{}: Art channel ready", host);
+                    stateMap.clear();
+                    if (remoteControllerWebSocket.callback.getArtMode2022()) {
+                        remoteControllerWebSocket.callback.setArtMode2022(false);
+                        remoteControllerWebSocket.callback.setArtModeSupported(true);
+                        logger.info("{}: Art Mode has been renabled on Frame TV's >= 2022", host);
```

Comment:
```suggestion
```

Reply:
If I took this line out, how would anyone know that this 2022+ Frame TV supports the art mode api? some do, and some don't. This logging is absolutely necessary.

Reply:
ok